### ISSUE TITLE
Public dashboard caching is slow with field filter parameters

### ISSUE THREAD

**Issue by speedfire514** [metabase/metabase#16171]
**Describe the bug**
We tried to leverage the cache using the public dashboard for some heavier requests (in that case in MongoDB but it also happens on SQL). Requests are parametrized using field filter without default value (no value will fetch the whole dataset)

In the case that we don't provide any value in the field filters in the public dashboard, the request will obviously download the full set of data  (~2sec), but all the others subsequent calls will be really fast (~200ms), because it is fetched from the cache.
The second I add a field filter value, the first un-cached call might take longer (~30sec) and every subsequent calls hitting the cache will still be slow to return data (~15 sec).
Timing are measure by monitoring calls in the browser consoles and also by the API GET completed call logs reporting the request time. I can confirm as well cache hitting by looking into the log.

The problem gets bigger the more values is filled to the board. We have a case of a list of 450s value for a field filter, leading to the 15s cache latency. If I reduced the field filter to 1 selected value, cache latency is 2s. If I remove the field filter and fetch more data, cache latency is 200ms.

All that information gathered I feel like there might be a performance issue serializing field filter.

Value filters does not seem to create that issue (200ms was using a value filter filled)

The mongo request was tested with all the 450s value and without to assert db latency. None of them is slow enough (~3sec to fetch)

After more testing : 

It appears that any value fed to filer containing '/' makes the cache to slowly respond, it did not depend on the type of filter as is as a field filter without slashed is fast. and a simple filter with a number is fast as well.

Is it a collation problem in the db ?

Update:
Narrowed down is the issue to the fact that such parameters in TRACE level makes calls too slow. I still don't understand how that kind of field value has an impact on that.

**Logs**

I activated TRACE/DEBUG and there is a lot of log not to mention that 450 value is populating the log quite heavily. I can narrow down to some namespace or specific section of the log if requested

**To Reproduce**
1. Create a heavy enough request with field filter (ie on mongo) to trigger caching
2. Maps a field filter without default value
3. Share the dashboard
4. Open the dashboard without any value and refresh to hit the cache
5. Repeat 4 but with a value selected in the field filter
6. Repeat 4 with a lot of value in the filed filter (easier to do using a copy-paste into the window)

**Expected behavior**
Cached query results should be faster hitting the cache. (filter should also reduce the data size to put and fetch from cache)

**Screenshots**

**Information about your Metabase Installation:**
```
{
  "browser-info": {
    "language": "en-US",
    "platform": "Linux x86_64",
    "userAgent": "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/90.0.4430.214 Safari/537.36",
    "vendor": "Google Inc."
  },
  "system-info": {
    "file.encoding": "UTF-8",
    "java.runtime.name": "OpenJDK Runtime Environment",
    "java.runtime.version": "11.0.11+9",
    "java.vendor": "AdoptOpenJDK",
    "java.vendor.url": "https://adoptopenjdk.net/",
    "java.version": "11.0.11",
    "java.vm.name": "OpenJDK 64-Bit Server VM",
    "java.vm.version": "11.0.11+9",
    "os.name": "Linux",
    "os.version": "5.4.0-1026-kvm",
    "user.language": "en",
    "user.timezone": "GMT"
  },
  "metabase-info": {
    "databases": [
      "sqlserver",
      "mongo"
    ],
    "hosting-env": "unknown",
    "application-database": "mysql",
    "application-database-details": {
      "database": {
        "name": "MySQL",
        "version": "5.7.12"
      },
      "jdbc-driver": {
        "name": "MariaDB Connector/J",
        "version": "2.6.2"
      }
    },
    "run-mode": "prod",
    "version": {
      "date": "2021-05-17",
      "tag": "v0.39.2",
      "branch": "release-x.39.x",
      "hash": "b0d9436"
    },
    "settings": {
      "report-timezone": null
    }
  }
}
```
- Vivaldi (chromium) 3.8.2259.42 /firefox and Edge latest
- Fedora 34
- Your databases: MySQL, Mongo
- Metabase version: 0.39.2 
- Metabase hosting environment: Docker on Kubernetes
- Metabase internal database: External AWS Aurora Mysql

**Severity**
Moderate : Increasing Ingress timeout to mitigate the problem. Users still complain about slow response of shared dashboards

**Additional context**
I can process different tests to narrow down the problem or provide part of log required for further investigation

**Comment by flamber** [comment]
Hi @speedfire514 

Very interesting write-up you have made.

Since you say you can reproduce the problem on non-Mongo, can we then use that instead? It's much easier for me to test things with most other databases.

How many rows are you returning? Are you seeing the same behavior for slow queries that are just returning a couple of rows?

I'm not able to reproduce, so it would be very useful to get the logs to have a better understanding of the timings.
And anything that can help me reproduce this would be very helpful.

You are noting collation as a possible culprit - I don't see how, but run this on your application database:
`SELECT @@character_set_database, @@collation_database`

In "Expected behavior" you write "filter should also reduce the data size to put and fetch from cache", but unless I'm misunderstanding what you're saying, that's not how Metabase cache functions.
The cache is stored by per-query (with a hash), so if you make a query without a filter, then it'll get hashA, and if you then set a filter=1, then that query will get hashB, and filter=2 would get hashC.
In other words, Metabase does not store a full resultset in cache and then when filters are used, it queries the cache. Each query is send to the data source and then stored in the cache if there's results.

**Comment by speedfire514** [comment]
Hi @flamber, 
thanks for taking the time to help us with that. First, I edited a couple of times that issue to update with our recent finding. We also read the source to better understand implied mechanics. Like I said, we narrowed down the main culprit part of the responsiveness of the API to the logging level being too verbose (it basically lists all our long filter values multiple times). On a side note, I was surprised by that behaviour as I thought the log buffer would be flushed on a different thread. We basically reduce the logging level to get an acceptable user experience which in my idea, reduce the criticality of that bug. That being said, we can still see that the public API using the cache is slower than the dashboard in "private". 

I will be able to produce the result that you requested and report specific timing. I think it is better to get the log with the trace level activated. Anyway, I'll gather all of that.

> You are noting collation as a possible culprit - I don't see how, but run this on your application database:
> `SELECT @@character_set_database, @@collation_database`

Yes I m not very convinced that it is related to the db collation either.  Part of it was because the '/' seemed to induce the problem.

> In "Expected behavior" you write "filter should also reduce the data size to put and fetch from cache", but unless I'm misunderstanding what you're saying, that's not how Metabase cache functions.
> The cache is stored by per-query (with a hash), so if you make a query without a filter, then it'll get hashA, and if you then set a filter=1, then that query will get hashB, and filter=2 would get hashC.
> In other words, Metabase does not store a full resultset in cache and then when filters are used, it queries the cache. Each query is send to the data source and then stored in the cache if there's results.

Yes, that's what I mean. I did not phrase it correctly. I meant that it is weird that a reduced set in cache (because of the query being filtered) took more time than fetching the full data set from the database.

**Comment by flamber** [comment]
@speedfire514 
When I initially read this issue, I instantly thought of #13150, but I was trying to get the logs from you to confirm.

I have spend an hour trying to reproduce, but I'm only seeing the problem 13150 outlines.

Can you reproduce with non-Mongo and provide the logs?

Run this on your application database? (with the same credentials as you're using for Metabase, selecting the database first) `SELECT @@character_set_database, @@collation_database`

Are you saying the the problem almost goes away, when you log less?

Like I said, I cannot reproduce and I need much more information to understand what's going on.

**Comment by flamber** [comment]
Closing, since without complete logs from Metabase and the database, then it's not possible to figure out what is going on. I cannot reproduce, besides the already referenced issue, but that's not specific to Public dashboards.
